File: tests/architecture-audit-checker/javascript/src/loader.js
/**
 * Deliverable loader for the architecture-audit checker harness (JavaScript track).
 *
 * Reads the six Markdown deliverables under `.kiro/specs/architecture-audit/`
 * into a {@link Deliverables} object. Missing files are tolerated and surface
 * as empty strings so downstream property tests can attribute the failure to a
 * specific deliverable instead of crashing on the load step.
 *
 * Pure ESM, no transpile step.
 */

import { readFile } from "node:fs/promises";
import path from "node:path";

import { makeDeliverables } from "./models.js";

/**
 * Closed map from logical deliverable name to its file basename. Mirrors
 * design.md §"Deliverable file locations (closed set)".
 *
 * @type {Readonly<Record<keyof import("./models.js").Deliverables, string>>}
 */
export const DELIVERABLE_FILENAMES = Object.freeze({
  audit_report: "audit-report.md",
  dependency_map: "dependency-map.md",
  migration_plan: "migration-plan.md",
  compatibility_strategy: "compatibility-strategy.md",
  preservation_inventory: "preservation-inventory.md",
  rebuild_recommendations: "rebuild-recommendations.md",
});

/**
 * Load all six deliverables from a spec directory.
 *
 * @param {string} specDir
 *   Absolute or workspace-relative path to the directory holding the six
 *   audit deliverables. Typically `.kiro/specs/architecture-audit/`.
 * @returns {Promise<import("./models.js").Deliverables>}
 *   A populated Deliverables object. Missing files yield empty strings rather
 *   than throwing, so the caller can decide whether absence is a property
 *   failure or expected for the test scenario.
 */
export async function loadDeliverables(specDir) {
  /** @type {import("./models.js").Deliverables} */
  const deliverables = makeDeliverables();

  await Promise.all(
    Object.entries(DELIVERABLE_FILENAMES).map(async ([key, filename]) => {
      const fullPath = path.join(specDir, filename);
      try {
        const content = await readFile(fullPath, "utf8");
        // @ts-ignore — key is statically one of the Deliverables fields
        deliverables[key] = content;
      } catch (err) {
        if (err && /** @type {NodeJS.ErrnoException} */ (err).code === "ENOENT") {
          // Tolerate missing deliverables; the parser layer surfaces this as
          // an empty document and the property tests can fail with a precise
          // attribution.
          // @ts-ignore — key is statically one of the Deliverables fields
          deliverables[key] = "";
          return;
        }
        throw err;
      }
    }),
  );

  return deliverables;
}

File: tests/architecture-audit-checker/rust/src/loader.rs
//! Read-only loader for the six Markdown deliverables under
//! `.kiro/specs/architecture-audit/`.
//!
//! The loader never writes to disk and never executes shell commands. A
//! missing deliverable file is represented as the empty string so the 17
//! Correctness Properties can still run and report structural defects rather
//! than crashing on `ENOENT`.

use std::fs;
use std::path::Path;

use crate::models::Deliverables;

/// Filename of each deliverable, in the canonical order documented in
/// `.kiro/specs/architecture-audit/design.md`.
const AUDIT_REPORT: &str = "audit-report.md";
const DEPENDENCY_MAP: &str = "dependency-map.md";
const MIGRATION_PLAN: &str = "migration-plan.md";
const COMPATIBILITY_STRATEGY: &str = "compatibility-strategy.md";
const PRESERVATION_INVENTORY: &str = "preservation-inventory.md";
const REBUILD_RECOMMENDATIONS: &str = "rebuild-recommendations.md";

/// Load every deliverable Markdown file from `spec_dir`. Files that do not
/// exist (or cannot be read) yield an empty string for that field.
pub fn load_deliverables(spec_dir: &Path) -> Deliverables {
    Deliverables {
        audit_report: read_or_empty(spec_dir, AUDIT_REPORT),
        dependency_map: read_or_empty(spec_dir, DEPENDENCY_MAP),
        migration_plan: read_or_empty(spec_dir, MIGRATION_PLAN),
        compatibility_strategy: read_or_empty(spec_dir, COMPATIBILITY_STRATEGY),
        preservation_inventory: read_or_empty(spec_dir, PRESERVATION_INVENTORY),
        rebuild_recommendations: read_or_empty(spec_dir, REBUILD_RECOMMENDATIONS),
    }
}

fn read_or_empty(spec_dir: &Path, filename: &str) -> String {
    fs::read_to_string(spec_dir.join(filename)).unwrap_or_default()
}
